arch/arm/core/aarch32/cortex_r/reset.S

```diff
@@ -26,77 +27,161 @@ GDATA(z_arm_sys_stack)
 GDATA(z_arm_fiq_stack)
 GDATA(z_arm_abort_stack)
 GDATA(z_arm_undef_stack)
-
-#define STACK_MARGIN	4
-
+#if defined(CONFIG_PLATFORM_SPECIFIC_INIT)
+GTEXT(z_platform_init)
+#endif
 
 /**
  *
  * @brief Reset vector
  *
- * Ran when the system comes out of reset. The processor is in thread mode with
- * privileged level. At this point, the main stack pointer (MSP) is already
- * pointing to a valid area in SRAM.
+ * Ran when the system comes out of reset. The processor is in Supervisor mode
+ * and interrupts are disabled. The processor architectural registers are in
+ * an indeterminate state.
  *
  * When these steps are completed, jump to z_arm_prep_c(), which will finish
  * setting up the system for running C code.
  *
  * @return N/A
  */
 SECTION_SUBSEC_FUNC(TEXT, _reset_section, z_arm_reset)
-SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
-	mov r0,  #0
-	mov r1,  #0
-	mov r2,  #0
-	mov r3,  #0
-	mov r4,  #0
-	mov r5,  #0
-	mov r6,  #0
-	mov r7,  #0
-	mov r8,  #0
-	mov r9,  #0
-	mov r10, #0
-	mov r11, #0
-	mov r12, #0
-	mov r14, #0
-
-	/* lock interrupts: will get unlocked when switch to main task */
-	cpsid if
-
-	/* Setup FIQ stack */
-	msr CPSR_c, #(MODE_FIQ | I_BIT | F_BIT)
-	ldr sp, =(z_arm_fiq_stack + CONFIG_ARMV7_FIQ_STACK_SIZE - STACK_MARGIN)
-
-	/* Setup IRQ stack */
-	msr CPSR_c, #(MODE_IRQ | I_BIT | F_BIT)
-	ldr sp, =(_interrupt_stack + CONFIG_ISR_STACK_SIZE - STACK_MARGIN)
-
-	/* Setup data abort stack */
-	msr CPSR_c, #(MODE_ABT | I_BIT | F_BIT)
-	ldr sp, =(z_arm_abort_stack + CONFIG_ARMV7_EXCEPTION_STACK_SIZE - \
-				STACK_MARGIN)
-
-	/* Setup undefined mode stack */
-	msr CPSR_c, #(MODE_UDF | I_BIT | F_BIT)
-	ldr sp, =(z_arm_undef_stack + CONFIG_ARMV7_EXCEPTION_STACK_SIZE - \
-				STACK_MARGIN)
-
-	/* Setup SVC mode stack */
-	msr CPSR_c, #(MODE_SVC | I_BIT | F_BIT)
-	ldr sp, =(z_arm_svc_stack + CONFIG_ARMV7_SVC_STACK_SIZE - STACK_MARGIN)
-
-	/* Setup System mode stack */
-	msr CPSR_c, #(MODE_SYS | I_BIT | F_BIT)
-	ldr sp, =(z_arm_sys_stack + CONFIG_ARMV7_SYS_STACK_SIZE - STACK_MARGIN)
-
-	/* Setup system control register */
-	mrc p15, 0, r0, c1, c0, 0	/* SCTLR */
-	bic r0, r0, #HIVECS		/* Exception vectors from 0-0x1c */
-	mcr p15, 0, r0, c1, c0, 0
+SECTION_SUBSEC_FUNC(TEXT, _reset_section, __start)
+
+#if defined(CONFIG_CPU_HAS_DCLS)
```

Comment:
This is a change from default behavior, i.e. when DCLS is not set. Could you comment why the clearing of common registers is not required in that case? Especially since you state that the processor architectural registers are in indeterminate state..

Reply:
> Could you comment why the clearing of common registers is not required in that case?

@ioannisg "Indeterminate register reset value" in itself is not really a problem from the program code perspective because all architectural registers are going to be, at some point, set to a known value before being used. Note that Cortex-A behaves the same way yet we do not initialise registers in the boot code because it is simply not necessary to do so.

The register initialisation sequence is necessary if and only if the Dual-redundant Core Lock-step topology is implemented, because the register values of the two redundant cores are not guaranteed to be identical due to the indeterminate nature of the register reset state, and this can cause the lock-step logic to detect it as a fault (think of it as "synchronisation"/"convergence" of the register states of the dual redundant cores).

```
    /*
     * Initialise CPU registers to a defined state if the processor is
     * configured as Dual-redundant Core Lock-step (DCLS). This is required
     * for state convergence of the two parallel executing cores.
     */
```

Reply:
Right,  I understood that; wanted to make sure you were aware of this update in the dflt behavior